### PR TITLE
Minor code cleanup of sherpa.astro.ui.utils

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

The `sherpa.astro.ui.utils` module has seen a number of minor clean-ups, addressing pylint-reported issues.

# Details

The aim was to address technical debt by applying pylint suggestions. Many of the lines changed were identified and tests added (and these have already been merged in #1067), so hopefully the coverage is high for these changes. Tests have been added for some of the un-tested changes.

This used to contain code that has been moved out to #1067, #1072, #1073

### CODE REVIEW

File: sherpa/astro/ui/utils.py [hunk 1]
```diff
@@ -3788,8 +3790,10 @@ def _save_type(self, objtype, id, filename, bkg_id=None, **kwargs):
         if bkg_id is not None:
             d = self.get_bkg(id, bkg_id)
 
-        if type(d) in (sherpa.astro.data.DataIMG, sherpa.astro.data.DataIMGInt,
-                       sherpa.data.Data2D, sherpa.data.Data2DInt):
+        if isinstance(d, (sherpa.astro.data.DataIMG,
```
Comment (hamogu) [inline sherpa/astro/ui/utils.py:3793]: Would it be easier (and more future proof) to simply test the number of axes, i.e. the spacial dimension of the independent variable?

**Review by hamogu — CHANGES_REQUESTED** [review]

**Comment by hamogu** [comment]
Again thank you for spending so much time to improve tests and the largely invisible, but absolutely necessary work of cleaning thing up.

File: sherpa/astro/ui/tests/test_astro_ui_unit.py [hunk 2]
```diff
@@ -1275,3 +1275,241 @@ def test_delete_bkg_model_with_bkgid(id, clean_astro_ui):
...
+
+    data = ui.get_data(idval)
+    assert data.grouping is None
+
+    if idval is None:
+        ui.load_grouping(str(path))
+    else:
+        ui.load_grouping(idval, str(path))
+
+    assert not data.grouped
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/ui/tests/test_astro_ui_unit.py:1333]: The documentation for `load_grouping` is explicit (but I, like you, had not realized it until my initial 'assert data.grouped' check failed).There is actually a reason for this, since it's really the combination of grouping and quality that defines the grouping, although I think we could argue that the quality values only affect `ignore_bad` calls.

You can think of it a bit like background subtraction - we don't automatically subtract the background, so we don't automatically group. However, this falls apart because we do automatically apply the grouping when the file is first read in (i.e. we get both quality and grouping set).

Long-answer-short - I think it's an edge case and we at least document the behavior, so I'm not in a hurry to change it. We can review it as part of the grouping and filtering work.

File: sherpa/astro/ui/tests/test_astro_ui_unit.py [hunk 3]
```diff
@@ -1275,3 +1275,241 @@ def test_delete_bkg_model_with_bkgid(id, clean_astro_ui):
...
+    assert len(ax) == 2
+    assert ax[0] == pytest.approx([2, 10, 20])
+    assert ax[1] == pytest.approx([10, 12, 22])
+
+
+def test_get_axes_datapha_no_response():
+    """Since we have no response this is a bit odd.
+
+    I have noted that it's unclear whether the bin edges are
+    1-2, 2-3, 3-4 or 0.5-1.5, 1.5-2.5, 2.5-3.5. Let's test
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/ui/tests/test_astro_ui_unit.py:1439]: I have some recent-ish discussion/PR where I have pointed this put. I think the 0.5-1.5, ... case is wrong and we should return 1-2, ... but that's a different PR!

File: sherpa/astro/ui/utils.py [hunk 4]
```diff
@@ -650,6 +647,8 @@ def dataspace1d(self, start, stop, step=1, numbins=None,
 
         """
         # support non-integrated grids with inclusive boundaries
+        # We do NOT want to use an isinstance check since Data1DInt is
+        # derived from Data1D.
         if dstype in (sherpa.data.Data1D, sherpa.astro.data.DataPHA):
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/ui/utils.py:652]: @olaurino had a question about this in his work cleaning up the Data hierarchy in #614 but I'm  not sure we ever resolved this. Technically the PHA data is integrated, but the PHA standard/documentation kind-of treats it as channel, counts (or rate) rather than channel_lo, channel_hi, counts and our code has (I fell) implicitly treated it more like a Data1D rather than Data1DInt class (but not always).

If you look at DataPHA https://sherpa.readthedocs.io/en/latest/data/api/sherpa.astro.data.DataPHA.html you'll see that it derives from Data1D (probably unfortunately).

File: sherpa/astro/ui/utils.py [hunk 5]
```diff
@@ -1438,28 +1435,23 @@ def load_ascii_with_errors(self, id, filename=None, colkeys=None, sep=' ',
...
+            data.elo = data.y - data.elo
+            data.ehi = data.ehi - data.y
 
-        if type(datas) is list:
-            for data in datas:
-                if delta is False:
-                    data.elo, data.ehi = calc_err(data)
-                data.staterror = calc_staterror(data)
+        if func is numpy.average:
+            staterror = func([data.elo, data.ehi], axis=0)
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/ui/utils.py:1446]: I think at this point we should be able to just access the attributes without data chackes, and rely on the setters for the data object to do the validation. Otherwise every place you use `data.elo` you need to add a check

File: sherpa/astro/ui/utils.py (hunk 1 above)
Comment (DougBurke) [reply sherpa/astro/ui/utils.py:3793]: This is a completely-undocumented function that I don't understand the pre- or post-conditions for use, or even what it is meant to be doing, so I am loath to change it significantly at this time!

File: sherpa/astro/ui/tests/test_astro_ui_unit.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hamogu) [reply sherpa/astro/ui/tests/test_astro_ui_unit.py:1333]: >  I'm not in a hurry to change it. We can review it as part of the grouping and filtering work.
Agreed.

File: sherpa/astro/ui/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hamogu) [reply sherpa/astro/ui/utils.py:1446]: OK.

**Review by hamogu — DISMISSED** [review]

**Comment by hamogu** [comment]
Except for the two extra assert statements that I suggested, everything else is fine with me!

**Review by hamogu — APPROVED** [review]

**Comment by DougBurke** [comment]
Hold on, something must have got borked with the auto-generated update. let me look...

**Comment by DougBurke** [comment]
So I wasn't careful enough with reviewing the suggested fixes for the indentation. This has been fixed now. Let's hope CI agrees.

**Comment by hamogu** [comment]
When I approve, I also mean "approved, but don't merge until CI passes" (unless there is a very good reason otherwise!) ;-)

**Comment by DougBurke** [comment]
Phew. Green.
